File: tusimple/递归/54.spiral_matrix.py
class Solution:
    def spiralOrder(self, matrix):
        res = []
        m, n = len(matrix), len(matrix[0])
        size = m * n
        visited = [[0 for _ in range(n)] for _ in range(m)]

        turn = {'right':'down', 'down':'left', 'left':'up', 'up':'right'}
        transfer = {'right': (0, 1), 'down': (1, 0), 'left': (0, -1), 'up': (-1, 0)}

        def helper(r, c, direction):
            res.append(matrix[r][c])
            visited[r][c] = 1
            if len(res) == size:
                return
            nr, nc = r+transfer[direction][0], c+transfer[direction][1]
            if 0<=nr<m and 0<=nc<n and not visited[nr][nc]:
                helper(nr, nc, direction)
            else:
                newDirection = turn[direction]
                nr, nc = r+transfer[newDirection][0], c+transfer[newDirection][1]
                helper(nr, nc, newDirection)

        helper(0, 0, 'right')
        return res

File: C++/Array/54. Spiral Matrix.cpp
class Solution {
public:
    
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> res;
        int m = matrix.size(), n = matrix[0].size();
        int len = m * n;
        int r = 0, c = 0;
        string next = "right";
        while (true) {
            res.push_back(matrix[r][c]); 
            if (res.size() == len) break;
            matrix[r][c] = 101; // 101是一个超出取值范围的数，表示已访问
            if (next == "right") {
                if (c+1<n && matrix[r][c+1]!=101) c++;
                else { next = "down"; r++; }
            }
            else if (next == "down") {
                if (r+1<m && matrix[r+1][c]!=101) r++;
                else { next = "left"; c--; }
            }
            else if (next == "left") {
                if (c-1>=0 && matrix[r][c-1]!=101) c--;
                else { next = "up"; r--; }
            }
            else {
                if (r-1>=0 && matrix[r-1][c]!=101) r--;
                else { next = "right"; c++; }
            }
        }
        return res;
    }
};